Source/Core/Core/PowerPC/TableGen.cpp

```diff
@@ -0,0 +1,459 @@
+  }
+  for (size_t i = 0; i < subtables.size(); i += 1)
+  {
+    CreateTable(table, decoding_table, decoding_table[offset + i], lines);
+  }
+}
+
+static void DoOutput(const std::vector<Instruction>& table,
+                     const std::vector<DecodingEntry>& decoding_table, const OutputOptions& options,
+                     std::ostream& out)
+{
+  if (options.flags & OPID_RANGES)
+  {
+    for (auto& dec : decoding_table)
+    {
+      out << dec.table_name << " = " << dec.instruction_offset + 1 << ", // " << dec.description
+          << "\n";
+      out << dec.table_name << "_End = " << dec.instruction_offset + 1 + dec.instructions.count()
+          << ",\n";
+    }
+    out << "End = " << table.size() + 1 << '\n';
+  }
+  if (options.flags & DECODING_TABLE)
+  {
+    for (auto& dec : decoding_table)
+    {
+      out << "{0x" << std::hex << std::setfill('0') << std::setw(16) << dec.instructions.to_ulong()
+          << ", 0x" << std::setw(16) << dec.subtables.to_ulong() << std::dec
+          << ", "
+          // for binary literals: "{0b" << dec.instructions << ", 0b" << dec.subtables
+          << dec.instruction_offset + 1 << ", " << dec.subtable_offset << ", " << dec.shift << ", "
+          << dec.length << "},\n";
+    }
+  }
+  if (options.flags & OPINFO_TABLE)
+  {
+    for (auto& inst : table)
+    {
+      out << "{\"" << inst.opname << "\", OpType::" << inst.type << ", " << inst.flags << "},\n";
+    }
+  }
+  if (!options.columns.empty())
+  {
+    for (auto& inst : table)
+    {
+      if (options.columns.size() > 1)
+      {
+        out << "{";
+      }
+      for (auto& col : options.columns)
+      {
+        const std::string* val;
+        if (inst.dispatch.size() > col.column && !inst.dispatch[col.column].empty())
+        {
+          val = &inst.dispatch[col.column];
+        }
+        else
+        {
+          val = &col.default_;
+        }
+        if (*val == "*")
+        {
+          out << col.prefix << inst.opname << ", \n";
+        }
+        else
+        {
+          out << *val << ", \n";
+        }
+      }
+      if (options.columns.size() > 1)
+      {
+        out << "}";
+      }
+    }
+  }
+}
+
+static void ReadTable(std::istream& in, std::vector<std::deque<std::string>>& rows)
+{
+  while (!in.eof())
+  {
+    std::string line;
+    std::getline(in, line);
+    std::deque<std::string>::size_type nextpos, pos = 0;
+    std::deque<std::string> row;
+    if (line.empty())
+    {
+      rows.push_back(row);
+      continue;
+    }
+    while (true)
+    {
+      nextpos = line.find(SEPARATOR, pos);
+      if (nextpos != std::string::npos)
+      {
+        row.push_back(line.substr(pos, nextpos - pos));
+        pos = nextpos + 1;
+      }
+      else
+      {
+        row.push_back(line.substr(pos, std::string::npos));
+        break;
+      }
+    }
+    rows.push_back(row);
+  }
+}
+
+int main(int argc, char** argv)
+{
+  std::vector<std::pair<std::ofstream, OutputOptions>> file_outputs;
```

Comment:
These should be moved closest to the scopes they're used in.

Reply:
I don't understand what you mean. This vector is filled by the command line parsing code.

Reply:
Oh geez, never mind that comment, I completely missed the ParseCommandLine() call.